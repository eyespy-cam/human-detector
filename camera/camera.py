class Camera:
    def __init__(self):
        self.camera = self.open_camera()

    def open_stream():
        return ""
    
    def open_camera():
        return None

    def get_frame():
        return None

def main():
    pass

if __file__ == "__main__":
    main()