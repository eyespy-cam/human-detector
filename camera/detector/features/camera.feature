Feature: Camera detection

    Scenario: Human is detected
        Given a human walks into frame
        When a human is detected
        Then a notification will be sent

    Scenario: Unauthorized cCamera
        Given a human walks into frame
        When a human is detected
        Then a notification fails to send

    Scenario: No human is detected
        Given a human is not in frame
        When a human is not detected
        Then a notification will not be sent
