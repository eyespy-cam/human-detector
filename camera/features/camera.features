# unauthrized camera
# human
# no human