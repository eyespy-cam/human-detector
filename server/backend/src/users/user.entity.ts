import {
  Collection,
  Entity,
  OneToMany,
  PrimaryKey,
  Property,
  UuidType,
} from '@mikro-orm/core';
import { Group } from '../groups/group.entity';
import { v4 } from 'uuid';

@Entity()
export class User {
  @PrimaryKey({ type: UuidType })
  id = v4();

  // A user might not have a push notification:
  // - After account registration
  // - If they never allow notification permissions in the app
  @Property({ nullable: true })
  expoToken?: string;

  @OneToMany(() => Group, (group) => group.user)
  groups = new Collection<Group>(this);
}
