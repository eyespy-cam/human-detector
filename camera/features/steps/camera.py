from behave import *
import cv2
import os
from src.detector import Detector
from eyeSocket import ServerConnection
from src.camera import Camera

@given('a human walks into frame')
def step_impl(ctx):
    dirname = os.path.realpath('.')
    ctx.frame = cv2.imread(os.path.join(dirname, "tests/images/human.jpg"))
    
@given('a human is not in frame')
def step_impl(ctx):
    dirname = os.path.realpath('.')
    ctx.frame = cv2.imread(os.path.join(dirname, "tests/images/no_human.jpg"))

@when('a human is detected')
def step_impl(ctx):
    detector = Detector()
    result = detector.detect(ctx.frame)
    assert len(result) == 1

@when('a human is not detected')
def step_impl(ctx):
    detector = Detector()
    result = detector.detect(ctx.frame)
    assert len(result) == 0

@then('a notification will be sent')
def step_impl(ctx):
    socket = ServerConnection("ABCD", "https://api.eyespy.com")
    result = socket.send_notification(ctx.frame)
    assert result.status_code == 200

@then('a notification will not be sent')
def step_impl(ctx):
    pass

@then('a notification fails to send')
def step_impl(ctx):
    socket = ServerConnection("AKJFD:LSJF", "https://api.eyespy.com")
    result = socket.send_notification(ctx.frame)
    assert result.status_code != 200
