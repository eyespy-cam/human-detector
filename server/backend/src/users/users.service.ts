import { Collection, EntityRepository } from '@mikro-orm/core';
import { InjectRepository } from '@mikro-orm/nestjs';
import { Injectable } from '@nestjs/common';
import { NotFoundError } from '../errors.types';
import { Camera } from '../cameras/camera.entity'
import { Group } from '../groups/group.entity';
import { User } from './user.entity';

@Injectable()
export class UsersService {
  constructor(
    @InjectRepository(User) private usersRepository: EntityRepository<User>,
  ) {}

  /**
   * Gets the given user's camera groups.
   *
   * @param userId the user's ID.
   * @throws {@link NotFoundError}
   * Thrown if the given user ID doesn't correspond to an existing user.
   */
  public async getGroups(userId: string): Promise<Collection<Group>> {
    const user = await this.usersRepository.findOne(
      { id: userId },
      { populate: ['groups.cameras'] },
    );
    if (user === null) {
      throw new NotFoundError(`User with ID "${userId}" does not exist.`);
    }
    return user.groups;
  }

  public async putCamera(userId: string, groupId: string, name: string, pubKey: string, serial: string): Promise<Camera> {
    
    return null;
  }

  /**
   * Update's the user's push notification token.
   *
   * @param userId the user's ID.
   * @param notifyToken the new notification token.
   */
  public async updateNotifyToken(
    userId: string,
    notifyToken: string,
  ): Promise<void> {
    const user = await this.usersRepository.findOne({ id: userId });
    if (user === null) {
      throw new NotFoundError(`User with ID "${userId}" does not exist.`);
    }
    user.expoToken = notifyToken;
    await this.usersRepository.flush();
  }
}
