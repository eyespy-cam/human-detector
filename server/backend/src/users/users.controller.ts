import {
  Body,
  Controller,
  ForbiddenException,
  Get,
  Inject,
  Param,
  Put,
  UseGuards,
} from '@nestjs/common';
import { JwtIdentityGuard } from '../auth/jwt-identity.guard';
import { NotFoundError } from '../errors.types';
import { UsersService } from './users.service';

export type GetGroupsOutput = {
  id: string;
  name: string;
  cameras: {
    id: string;
    name: string;
  }[];
}[];

export type PutCameraResponse = {
  id: string;
}

@Controller('users')
@UseGuards(JwtIdentityGuard)
export class UsersController {
  constructor(@Inject(UsersService) private usersService: UsersService) {}

  @Get(':id/groups')
  async getGroups(@Param('id') id: string): Promise<GetGroupsOutput> {
    try {
      const groupsJSON = (await this.usersService.getGroups(id)).toJSON();
      return groupsJSON.map((group) => ({
        id: group.id,
        name: group.name,
        cameras: group.cameras.map((camera) => ({
          id: camera.id,
          name: camera.name,
        })),
      }));
    } catch (error) {
      if (error instanceof NotFoundError) {
        throw new ForbiddenException();
      } else {
        throw error;
      }
    }
  }

  @Put(':uid/groups/:gid/cameras')
  async putCamera(
    @Param('uid') userId: string,
    @Param('gid') groupId: string,
    @Body('publicKey') publicKey: string,
    @Body('serial') cameraSerial: string,
    @Body('name') cameraName: string
  ): Promise<PutCameraResponse> {
    return null;
  }

  @Put(':id/notifyToken')
  async updateNotifyToken(
    @Param('id') id: string,
    @Body('expoToken') notifyToken: string,
  ): Promise<void> {
    try {
      await this.usersService.updateNotifyToken(id, notifyToken);
    } catch (error) {
      if (error instanceof NotFoundError) {
        throw new ForbiddenException();
      } else {
        throw error;
      }
    }
  }
}
