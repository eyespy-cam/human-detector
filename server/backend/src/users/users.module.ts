import { MikroOrmModule } from '@mikro-orm/nestjs';
import { Module } from '@nestjs/common';
import { Group } from '../groups/group.entity';
import { User } from './user.entity';
import { UsersController } from './users.controller';
import { UsersService } from './users.service';

@Module({
  imports: [MikroOrmModule.forFeature([User, Group])],
  controllers: [UsersController],
  providers: [UsersService],
})
export class UsersModule {}
