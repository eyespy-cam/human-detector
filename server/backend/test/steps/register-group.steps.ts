import { defineFeature, loadFeature } from 'jest-cucumber';
import { INestApplication } from '@nestjs/common';
import { v4 } from 'uuid';
import request from 'supertest';
import { EntityRepository, MikroORM } from '@mikro-orm/core';
import {
  buildTestStack,
  TestStack,
  TEST_STACK_TIMEOUT,
} from '../helpers/test-stack';
import { UsersModule } from '../../src/users/users.module';
import { User } from '../../src/users/user.entity';
import { Group } from '../../src/groups/group.entity';
import { Camera } from '../../src/cameras/camera.entity';

defineFeature(loadFeature('test/features/register-group.feature'), (test) => {
  let app: INestApplication;
  let testStack: TestStack;
  let usersRepository: EntityRepository<User>;
  let groupRepository: EntityRepository<Group>;
  let cameraRepository: EntityRepository<Camera>;

  beforeAll(async () => {
    testStack = await buildTestStack({ imports: [UsersModule] });
    usersRepository = testStack.module
      .get<MikroORM>(MikroORM)
      .em.getRepository(User);

    groupRepository = testStack.module
      .get<MikroORM>(MikroORM)
      .em.getRepository(Group);

    cameraRepository = testStack.module
      .get<MikroORM>(MikroORM)
      .em.getRepository(Camera);

    app = testStack.module.createNestApplication();
    await app.init();
  }, TEST_STACK_TIMEOUT);

  afterAll(async () => {
    await app.close();
    await testStack.kcContainer.stop();
    await testStack.dbContainer.stop();
  });

  test('User is registering a group from the app', ({
    given,
    when,
    then,
    and,
  }) => {
    let group: Group;
    let user: User;
    let token: string;
    let response: request.Response;
    let groupId: string;

    given('I have credentials', async () => {
      const { id, tokenSet } =
        await testStack.kcContainer.createDummyUserAndLogIn('users');
      user = await usersRepository.findOneOrFail({ id });
      group = new Group('Group-A');
      user.groups.add(group);
      await usersRepository.flush();
      token = tokenSet.access_token;
    });
    when('I register a group through the app', async () => {
      response = await request(app.getHttpServer())
        .put(`/users/${user.id}/group`)
        .send({
          name: 'New-Group',
        })
        .auth(token, { type: 'bearer' });
    });
    then('I receive the group ID', async () => {
      expect(response.statusCode).toBe(200);
      expect(response.type).toEqual('application/json');
      expect(response.body.id).toBeDefined();

      groupId = response.body.id;
      expect(
        await groupRepository.findOneOrFail({ id: groupId }),
      ).toBeInstanceOf(Group);
    });
    console.log(user);

    and('The group is registered', async () => {
      expect(user.groups).toHaveLength(2);
    });
  });

  test('User A is attempting to register a group to User B', ({
    given,
    when,
    then,
  }) => {
    let userB: User;
    let userTokenA: string;
    let response: request.Response;
    given("I have user A's credentials", async () => {
      const userBLogin = await testStack.kcContainer.createDummyUserAndLogIn(
        'users',
      );
      userB = await usersRepository.findOneOrFail({ id: userBLogin.id });
      const userALogin = await testStack.kcContainer.createDummyUserAndLogIn(
        'users',
      );
      userTokenA = userALogin.tokenSet.access_token;
      await usersRepository.flush();
    });

    when('I register the group to user B', async () => {
      response = await request(app.getHttpServer())
        .put(`/users/${userB.id}/group`)
        .send({ name: 'New-Group-A' })
        .auth(userTokenA, { type: 'bearer' });
    });

    then('I receive an unauthorized error', () => {
      expect(response.statusCode).toBe(403);
      expect(response.body).toMatchSnapshot();
    });
  });

  test('User A is attempting to register a group to a non-existant User B', ({
    given,
    when,
    then,
  }) => {
    const userIdB = v4();
    let userTokenA: string;
    let response: request.Response;

    given("I have user A's credentials", async () => {
      const { tokenSet } = await testStack.kcContainer.createDummyUserAndLogIn(
        'users',
      );
      userTokenA = tokenSet.access_token;
    });

    when('I register the group to user B', async () => {
      response = await request(app.getHttpServer())
        .put(`/users/${userIdB}/group`)
        .send({ name: 'New-Group-A' })
        .auth(userTokenA, { type: 'bearer' });
    });

    then('I receive an unauthorized error', () => {
      expect(response.statusCode).toBe(403);
      expect(response.body).toMatchSnapshot();
    });
  });
});
