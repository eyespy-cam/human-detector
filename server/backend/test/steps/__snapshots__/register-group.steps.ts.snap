// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register a group to a user User A is attempting to register a group to User B 1`] = `
Object {
  "error": "Forbidden",
  "message": "Forbidden resource",
  "statusCode": 403,
}
`;

exports[`Register a group to a user User A is attempting to register a group to a non-existant User B 1`] = `
Object {
  "error": "Forbidden",
  "message": "Forbidden resource",
  "statusCode": 403,
}
`;
